Document keycloak decision (#85)
vshn/application-catalog-docs: docs/modules/ROOT/pages/explanations/decisions/keycloak.adoc

= VSHN Managed Keycloak

== Problem

We need to provide Keycloak on Kubernetes with the following features:

> * Standalone and Cluster functionality (HA))

Reviewer: ```suggestion
* Standalone and Cluster functionality (HA)
```

* Ability to customize Keycloak (themes, providers, env variables)
* Regular Maintenance
* Version Upgrades

> For custom themes and providers we need to be able to start an initContainer that copies the themes and providers to the volume mount used by the keycloak container.

Reviewer: ```suggestion
For custom themes and providers we need to be able to start an initContainer that copies the themes and providers to the volume mount used by the Keycloak container.
```

== Solutions

> For Keycloak there are multiple operators as well as helm charts available.

Reviewer: ```suggestion
For Keycloak there are multiple operators as well as Helm Charts available.
```

The following section contains the solutions that have been looked at.

[cols="1,1,1,1,1"]
|===
|Requirements |https://www.keycloak.org/guides#operator[Official Operator] |https://glasskube.eu/docs/crd-reference/keycloak/[Glasskube Operator] |https://github.com/codecentric/helm-charts/tree/master/charts/keycloak[Codecentric Helmchart] |https://github.com/bitnami/charts/tree/main/bitnami/keycloak[Bitnami Helmchart]

|Standalone and cluster| ✅ | ❌ only standalone | ✅ | ✅

|Customize Keycloak Settings | ✅ | ❌ | ✅ | ✅

|Metrics | ❌ | ❌ | ✅ | ✅

|Use custom image | ✅ | ❌ | ✅ | ✅

|Regular Maintenance | ❌ | ❌ | ❌ | ❌

|Version Upgrades |✅ | ❌ | ✅ | ✅

|===

Some additional notes:

> The Glasskube operator supports various tools and the keycloak support is currently very minimalistic (see offical docs).

Reviewer: ```suggestion
The Glasskube operator supports various tools and the Keycloak support is currently very minimalistic (see https://glasskube.eu/docs/crd-reference/keycloak/[official docs^]).
```

> The official keycloak operator officially only supports a handful of settings. However, it has been designed in a way to completely customize keycloak.

Reviewer: ```suggestion
The official Keycloak operator officially only supports a handful of settings. However, it has been designed in a way to completely customize Keycloak.
```

A bit too much "official" =)

Author: True. Removed one of them :-)

> It uses deployments for deploying keycloak.

Reviewer: ```suggestion
It uses deployments for deploying Keycloak.
```

> The codecentric helm chart is very flexible and supports a lot of settings. It uses a statefulset for deploying keycloak.

Reviewer: ```suggestion
The Codecentric Helm Chart is very flexible and supports a lot of settings. It uses a StatefulSet for deploying Keycloak.
```

> The bitnami helm chart has the most settings and flexibility. However, the chart works best in conjunction with the bitnami keycloak image and certain settings might not work correctly if using a different image.

Reviewer: ```suggestion
The Bitnami Helm Chart has the most settings and flexibility. However, the Chart works best in conjunction with the Bitnami Keycloak image and certain settings might not work correctly if using a different image.
```

== Decision

https://github.com/codecentric/helm-charts/tree/master/charts/keycloak[Codecentric Helmchart].

The instantiation of the Helm Charts will be handled by `+provider-kubernetes+`.

Advantages::

* Less complexity than operators, but provides the same features

> * We have been using this helm chart to deploy keycloak for years. We know it works.

Reviewer: ```suggestion
* We have been using this helm chart to deploy Keycloak for years. We know it works.
```

> * The helm chart is very flexible and supports a lot of settings
> * The helm chart is actively maintained and updated regularly

Reviewer: ```suggestion
* The Helm Chart is very flexible and supports a lot of settings
* The Helm Chart is actively maintained and updated regularly
```

> * No need to create a provider for a keycloak operator.

Reviewer: ```suggestion
* No need to create a provider for a Keycloak operator.
```

> * Migrating to another helm chart or forking the helm chart if neeeded is easier than replacing an operator.

Reviewer: ```suggestion
* Migrating to another Helm Chart or forking the Helm Chart if neeeded is easier than replacing an operator.
```

Disadvantages::

* Lack of regular, automated maintenance

> === Rational

Reviewer: ```suggestion
=== Rationale
```

Both operators are not very mature and lack basic features.

Although it is possible to configure every aspect of the deployment with the official operator using the `+unsupported+` field. It is officially not supported by Keycloak.

The Glasskube operator is very rudimentary and does not support any of the features we need.

The official operator is written in Java and the Glasskube operator is written in Kotlin.

> We lack the expertise to contribute to either of these projects.

Reviewer 2: I would not say we lack the expertise to contribute there, at least for Java. We should probably point to the fact that the current available operators are not customizable enough for our needs.

Author: Adjusted it accordingly that we only lack know-how in Kotlin.
the fact that the operators are not customizable enough is already documented.

> As we already have experience deploying services with helm charts and we have been using the codecentric helm chart for years, we are confident that this is the best solution for us.

Reviewer: ```suggestion
As we already have experience deploying services with Helm Charts and we have been using the Codecentric Helm Chart for years, we are confident that this is the best solution for us.
```

> Furthermore we don't need to write a custom provider as we can use the `+provider-kubernetes+` provider to instantiate the helm chart.

Reviewer: ```suggestion
Furthermore we don't need to write a custom provider as we can use the `+provider-kubernetes+` provider to instantiate the Helm Chart.
```

Why not use `provider-helm`? Or is this a typo?

Author: It's a typo. Fixed
